source/common/tracing/http_tracer_impl.cc

```diff
@@ -22,22 +22,53 @@
 #include "source/common/protobuf/utility.h"
 #include "source/common/stream_info/utility.h"
 
-#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
 
 namespace Envoy {
 namespace Tracing {
 
-// TODO(perf): Avoid string creations/copies in this entire file.
-static std::string buildResponseCode(const StreamInfo::StreamInfo& info) {
-  return info.responseCode() ? std::to_string(info.responseCode().value()) : "0";
+// Pre-allocate common response codes to minimize string creations.
+namespace {
+constexpr absl::string_view HttpResponseCode0 = "0";
+constexpr absl::string_view HttpResponseCode200 = "200";
+constexpr absl::string_view HttpResponseCode404 = "404";
+constexpr absl::string_view HttpResponseCode500 = "500";
+constexpr absl::string_view HttpResponseCode502 = "502";
+constexpr absl::string_view HttpResponseCode503 = "503";
+constexpr absl::string_view DefaultValue = "-";
+} // namespace
+
+// Helper for building response code string without excess allocations.
+static absl::string_view buildResponseCode(const StreamInfo::StreamInfo& info,
+                                           std::string& out_buffer) {
+  if (!info.responseCode()) {
+    return HttpResponseCode0;
+  }
+  const uint16_t code = info.responseCode().value();
+  switch (code) {
+  case 200:
+    return HttpResponseCode200;
+  case 404:
+    return HttpResponseCode404;
+  case 500:
+    return HttpResponseCode500;
+  case 502:
+    return HttpResponseCode502;
+  case 503:
+    return HttpResponseCode503;
+  default:
+    // Only allocate if code is uncommon
+    out_buffer = std::to_string(code);
+    return out_buffer;
+  }
```

Comment:
Would like to see a benchmark to justify this. I would expect std::string wouldn't actually do any allocations during this because std::string has an internal buffer for short strings, and these are all short strings.
([Existing conversation](https://groups.google.com/a/isocpp.org/g/std-proposals/c/OHF7J9jU1gc) on this topic.)

Reply:
Thanks @ravenblackx for all the thoughts! Yes, agree, I was doing some micro [benchmarks](https://quick-bench.com/q/mAij8t2A4fFnQT7c-ULF6C5zxHg) in quick-bench, and the multiple branch with miss is indeed much worse. Changed to the `if else` approach with only `0` and `200`.

Reply:
`StreamInfo` arg needs namespacing.

Reply:
(Also interesting is how extremely differently these benchmarks play out if you change which STL is in use - not calling `to_string` is way more valuable against `libc++`. And to *really* benchmark it you'd want to have like 80% 200s and 20% mixed other things in a single benchmark run, to account for CPU predictive adaptation, haha.)

Reply:
yea, I’ve seen `std::to_string` hit significantly harder on libc++ too. I think it mainly comes down to how `libc++` handles small string optimization and heap allocations. And you are right for the prediction set-up.